src/easylink/pipeline_graph.py

```diff
@@ -47,53 +50,81 @@ def update_slot_filepaths(self, config: Config) -> None:
+                    self[source][sink][edge_idx]["filepaths"] = (
                         str(
                             Path("intermediate")
                             / node
                             / imp_outputs[edge_attrs["output_slot"].name]
-                        )
-                    ]
+                        ),
+                    )
 
-    def get_input_slots(self, node: str) -> Dict[str, List[str]]:
+    def get_input_slots(self, node: str) -> dict[str, dict[str, Union[str, list[str]]]]:
         """Get all of a node's input slots from edges."""
-        input_slots = {}
-        for _, _, edge_attrs in self.in_edges(node, data=True):
-            # Consider whether we need duplicate variables to merge
-            env_var, files = edge_attrs["input_slot"].env_var, edge_attrs["filepaths"]
-            if env_var in input_slots:
-                input_slots[env_var].extend(files)
+        input_slots = [
+            edge_attrs["input_slot"] for _, _, edge_attrs in self.in_edges(node, data=True)
+        ]
+        filepaths_by_slot = [
+            list(edge_attrs["filepaths"])
+            for _, _, edge_attrs in self.in_edges(node, data=True)
+        ]
+        return self.condense_input_slots(input_slots, filepaths_by_slot)
+
+    @staticmethod
+    def condense_input_slots(
+        input_slots: List[InputSlot], filepaths_by_slot: List[str]
+    ) -> Dict[str, dict[str, Union[str, list[str]]]]:
+        condensed_slot_dict = {}
+        for input_slot, filepaths in zip(input_slots, filepaths_by_slot):
+            slot_name, env_var, validator = (
+                input_slot.name,
+                input_slot.env_var,
+                input_slot.validator,
+            )
+            if slot_name in condensed_slot_dict:
+                if env_var != condensed_slot_dict[slot_name]["env_var"]:
+                    raise ValueError(
+                        f"Duplicate slot name {slot_name} with different env vars."
+                    )
+                if validator != condensed_slot_dict[slot_name]["validator"]:
+                    raise ValueError(
+                        f"Duplicate slot name {slot_name} with different validators."
+                    )
+                condensed_slot_dict[slot_name]["filepaths"].extend(filepaths)
             else:
-                input_slots[env_var] = files
-        return input_slots
+                condensed_slot_dict[slot_name] = {
+                    "env_var": env_var,
+                    "validator": validator,
+                    "filepaths": filepaths,
+                }
+        return condensed_slot_dict
 
     def get_input_output_files(self, node: str) -> Tuple[List[str], List[str]]:
         """Get all of a node's input and output files from edges."""
         input_files = list(
             itertools.chain.from_iterable(
                 [
-                    edge_attrs["filepaths"]
+                    list(edge_attrs["filepaths"])
```

Comment:
I'm guessing these need to be lists for some reason?

Reply:
I was just using it as an extra primitive `copy` but I think it's probably unnecessary.